opacus/tests/ddp_hook_test.py

```diff
@@ -0,0 +1,278 @@
+
+    cleanup()
+
+
+def demo_ddp_hook(rank, world_size, weight, dp, noise_multiplier, max_grad_norm):
+    torch.manual_seed(rank)
+    batch_size = 32
+    withdp = "with" + ("out " if not dp else "")
+    print(f"Running DDP hook {withdp} differential privacy example on rank {rank}.")
+
+    device = setup_and_get_device(rank, world_size, nonce=1)
+
+    # create model and move it to GPU with id rank
+    model = ToyModel().to(device)
+    # print(f"Initial weight: {model.net1.weight.data}")
+
+    model.net1.bias.requires_grad = False
+    model.net2.bias.requires_grad = False
+    model.net2.weight.requires_grad = False
+
+    ddp_model = DDP(model, device_ids=[device])
+    # print(f"Initial weight after DDP wrapping: {ddp_model.module.net1.weight.data}")
+
+    if dp:
+        engine = PrivacyEngine(
+            ddp_model,
+            batch_size=batch_size,
+            sample_size=10 * batch_size,
+            alphas=PRIVACY_ALPHAS,
+            noise_multiplier=noise_multiplier,
+            max_grad_norm=max_grad_norm,
+            experimental=True,
+            clip_per_layer=True,
+        )
+        engine.random_number_generator = engine._set_seed(0)
+
+    loss_fn = nn.MSELoss()
+    optimizer = optim.SGD(ddp_model.parameters(), lr=1)
+    if dp:
+        engine.attach(optimizer)
+
+    optimizer.zero_grad()
+    labels = torch.randn(batch_size, 5).to(device)
+
+    outputs = ddp_model(torch.randn(batch_size, 10).to(device))
+    loss_fn(outputs, labels).backward()
+    optimizer.step()
+
+    # if rank == 0:
+    #     print(model.net1.weight)
+
+    weight.copy_(model.net1.weight.data.cpu())
+
+    del ddp_model
+    cleanup()
+
+
+def debug(rank, world_size, tensor, dp, noise_multiplier=0, max_grad_norm=1e8):
+    local_rank = setup_and_get_device(rank, world_size)
+    print(f"Rank: {rank},World size: {world_size}, local_rank: {local_rank}")
+    tensor = tensor.to(local_rank)
+    print(f"dp: {dp}")
+    print(tensor)
+
+    cleanup()
+
+
+def run_function(local_function, tensor, dp, noise_multiplier=0, max_grad_norm=1e8):
+    if os.environ.get("SLURM_NTASKS") is not None:
+        world_size = int(os.environ.get("SLURM_NTASKS"))
+        rank = int(os.environ.get("SLURM_PROCID"))
+        print(f"Running on a Slurm cluster with {world_size} tasks.")
+
+        local_function(rank, world_size, tensor, dp, noise_multiplier, max_grad_norm)
+    else:
+        world_size = torch.cuda.device_count()
+        print(f"Spawning multiple processes on a local machine with {world_size} GPUs")
+
+        # The rank will be passed as the first argument
+        mp.spawn(
+            local_function,
+            args=(
+                world_size,
+                tensor,
+                dp,
+                noise_multiplier,
+                max_grad_norm,
+            ),
+            nprocs=world_size,
+            join=True,
+        )
+    return world_size
+
+
+class GradientComputationTest(unittest.TestCase):
```

Comment:
Since you've essentially re-implemented clipping and adding noise, can you add specific tests for this too? Or instead update existing `per_sample_gradient_clip_test.py` and `privacy_engine_test.py` to include distributed case.

Reply:
Will do! I'll also try to reuse more code if possible